### PR TITLE
Switch `crypto.spec.ts` away from `TestClient` and `matrix-mock-request`.

### PR DESCRIPTION
**Pull request by richvdh**
First of all: sorry about the size of this PR. I've attempted to break it up into reviewable commits, but in the end we just have to do a big switcheroo.

I became sufficiently annoyed with `matrix-mock-request` that I decided to replace it with `fetch-mock`, which is what we use in `matrix-react-sdk` and is generally more powerful, easier to use, and actually maintained.

Unfortunately, we have a `TestClient` utility which is widely used and quite tightly integrated with `matrix-mock-request`. It wasn't going to be possible to change `TestClient` without changing all the tests that use it.

I also don't find `TestClient` particularly easy to use - it does a lot of stuff which I'm not convinced ought to be done for every single test.

So... I've introduced a couple of new classes (`SyncResponder`, `E2EKeyReceiver`) which do some of the useful bits of `TestClient`, but in a more granular way, and have switched `crypto.spec.ts` over so that rather than instantiating a `TestClient` for each test, it creates a `MatrixClient` directly and intercepts the endpoints necessary.

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

**Review by alunturner — APPROVED** [review]
Only one minor comment, I think this looks good to me though.

File: spec/test-utils/E2EKeyReceiver.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+     *
+     * If device keys have not yet been uploaded, throws an error
+     */
+    public getSigningKey(): string {
+        if (!this.deviceKeys) {
+            throw new Error("Device keys not yet uploaded");
+        }
+        const keyIds = Object.keys(this.deviceKeys.keys).filter((v) => v.startsWith("ed25519:"));
+        if (keyIds.length != 1) {
+            throw new Error(`Expected exactly 1 ed25519 key uploaded, got ${keyIds}`);
```
[the remark this answers is not in the document]
Comment (richvdh) [reply spec/test-utils/E2EKeyReceiver.ts:128]: it was supposed to be a stringfied array. I figured that would be more useful than just the number.

**Comment by richvdh** [comment]
Thanks for the review @alunturner !

**Review by andybalaam — APPROVED** [review]
Changes to package.json look innocuous, and the change is welcome generally. Thanks!

**Pull request merged**